File: MinStack1.java
// Time Complexity : O(1) for all operations
// Space Complexity : O(n) where n is number elements of hashCode. We used another stack to maintain min value with 1:1 corresponds
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
import java.util.Stack;

class MinStack1 {

    Stack<Integer> dataStack, minStack;
    int min;

    public MinStack1() {
        dataStack = new Stack();
        minStack = new Stack();
        min = Integer.MAX_VALUE;
    }

    public void push(int val) {
        min = Math.min(val, min);
        dataStack.push(val);
        minStack.push(min);
    }

    public void pop() {
        dataStack.pop();
        minStack.pop();
        if (minStack.isEmpty()) {
            min = Integer.MAX_VALUE;
        } else {
            min = minStack.peek();
        }
    }

    public int top() {
        return dataStack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}

File: minStack1.cpp
// Time Complexity : O(1)
// Space Complexity : O(2n) worst case, amortized O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
#include <climits>
class MinStack {
private:
    struct Node
    {
        int value;
        Node* next;
        Node(int val,Node* nextNode)
        {
            value = val;
            next = nextNode;
        }
    };
public:
    Node* head;
    int currMin;
    MinStack() {
        head = nullptr; 
        currMin = INT_MAX;
    }
    
    void push(int val) {
        if(val<=currMin)
        {
            head= new Node(currMin,head);
            currMin = val;
        }
        head = new Node(val,head);
    }
    
    void pop() {
        if(currMin==top())
        {
            Node* temp1 = head;
            Node* temp2 = head->next;
            currMin = temp2->value;
            head = head->next->next;
            delete temp1,temp2;
            return;
        }
        Node* temp = head;
        head = head->next;
        delete temp;
        return;
    }
    
    int top() {
        return head->value;
    }
    
    int getMin() {
        return currMin;
    }
};